### PR TITLE
[GEN] Add cache controls decoration to `2d_block_read` builtin calls

### PR DESCRIPTION
**Pull request by victor-eds**
#1224 converts `triton_gen.2Dblockload` to OCL builtin `2d_block_read` calls. Add cache controls decoration capturing original `cache_control` semantics.

### CODE REVIEW

**Comment by victor-eds** [comment]
Addresses #1233

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp [hunk 1]
```diff
@@ -245,12 +245,60 @@ static bool isOCLBuiltinAvailable(TritonGEN::Matrix2DBlockLoadOp op) {
...
+loadCacheControlToDecoration(Builder &builder, uint32_t operandNum,
+                             TritonGEN::LoadCacheControl orig) {
+  const auto build = [&builder,
+                      operandNum](TritonGEN::LoadCacheControlDecorationEnum l1,
+                                  TritonGEN::LoadCacheControlDecorationEnum l3)
+      -> SmallVector<Attribute> {
+    return {builder.getAttr<TritonGEN::LoadCacheControlDecorationAttr>(
+                1, l1, operandNum),
+            builder.getAttr<TritonGEN::LoadCacheControlDecorationAttr>(
+                3, l3, operandNum)};
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:261]: on PVC, SPIRV cache level 0 maps to L1 and cache level 1 maps to L3.
```suggestion
                0, l1, operandNum),
            builder.getAttr<TritonGEN::LoadCacheControlDecorationAttr>(
                1, l3, operandNum)};
```

**Comment by whitneywhtsang** [comment]
Can you please check if the expected GenISA with expected cache control is generated at the end from shader dump?

**Comment by victor-eds** [comment]
> Can you please check if the expected GenISA with expected cache control is generated at the end from shader dump?

Issue's this won't work till IGC picks https://github.com/KhronosGroup/SPIRV-LLVM-Translator/commit/c8bfc33ef8f9f273546a49d87a6612de0791e3d4.

**Comment by whitneywhtsang** [comment]
> > Can you please check if the expected GenISA with expected cache control is generated at the end from shader dump?
> 
> Issue's this won't work till IGC picks [KhronosGroup/SPIRV-LLVM-Translator@c8bfc33](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/commit/c8bfc33ef8f9f273546a49d87a6612de0791e3d4).

IIUC the feature is needed for LLVMIR -> SPV translation, which doesn't need IGC to update the SPV -> LLVMIR portion. https://github.com/intel/intel-xpu-backend-for-triton/pull/1251 updated the SPIRV-LLVM-Translator Triton uses.

**Comment by victor-eds** [comment]
[ola.txt](https://github.com/user-attachments/files/15593096/ola.txt)

Here's the full SPIR-V dump. It's `.txt` as apparently GitHub cares about extensions :facepalm:

File: test/TritonGEN/tritongen-to-llvm.mlir [hunk 2]
```diff
@@ -367,16 +367,152 @@ llvm.func @triton_gen.dpas.f32(%c : vector<8xf32>, %a : vector<4xf32>, %b : vect
...
+
+llvm.func @triton_gen.2Dblockload(%ptr : !llvm.ptr<1>, %base_width : i32, %base_height : i32, %base_pitch : i32, %x : i32, %y : i32) {
+  // CHECK:     llvm.func @triton_gen.2Dblockload(%arg0: !llvm.ptr<1>, %arg1: i32, %arg2: i32, %arg3: i32, %arg4: i32, %arg5: i32) {
+  // CHECK:  [[SIXTY:%.*]] = llvm.mlir.constant(16 : i32) : i32
+  // CHECK-NEXT:  [[DEST:%.*]] = llvm.alloca [[SIXTY]] x i16 : (i32) -> !llvm.ptr
+  // CHECK-DAG:  [[ZERO:%.*]] = llvm.mlir.constant(0 : i32) : i32
+  // CHECK-DAG:  [[ONE:%.*]] = llvm.mlir.constant(1 : i32) : i32
+  // CHECK-DAG:  [[UNDEF:%.*]] = llvm.mlir.undef : vector<2xi32>
+  // CHECK-NEXT: [[COORD0:%.*]] = llvm.insertelement %arg4, [[UNDEF]][[[ZERO]] : i32] : vector<2xi32>
+  // CHECK-NEXT: [[COORD1:%.*]] = llvm.insertelement %arg5, [[COORD0]][[[ONE]] : i32] : vector<2xi32>
```
[the remark this answers is not in the document]
Comment (victor-eds) [reply test/TritonGEN/tritongen-to-llvm.mlir:418]: I'll request for a change in the target PR to get better codegen and further reduce before applying this suggestion

File: test/TritonGEN/tritongen-to-llvm.mlir [hunk 3]
```diff
@@ -367,16 +367,152 @@ llvm.func @triton_gen.dpas.f32(%c : vector<8xf32>, %a : vector<4xf32>, %b : vect
 
 // -----
 
-// CHECK: llvm.func spir_funccc @intel_subgroup_block_read_u8_m8k32v2(!llvm.ptr<1>, i32, i32, i32, vector<2xi32>) -> vector<16xi16> attributes {passthrough = ["convergent"]}
+// CHECK: llvm.func spir_funccc @_Z40intel_sub_group_2d_block_read_8b_8r32x2cPU3AS1viiiDv2_iPt(!llvm.ptr<1>, i32, i32, i32, vector<2xi32>, !llvm.ptr) attributes {passthrough = ["convergent"]}
 
 llvm.func @triton_gen.2Dblockload(%ptr : !llvm.ptr<1>, %base_width : i32, %base_height : i32, %base_pitch : i32, %x : i32, %y : i32) {
   // CHECK:     llvm.func @triton_gen.2Dblockload(%arg0: !llvm.ptr<1>, %arg1: i32, %arg2: i32, %arg3: i32, %arg4: i32, %arg5: i32) {
+  // CHECK:  [[SIXTY:%.*]] = llvm.mlir.constant(16 : i32) : i32
```
[the remark this answers is not in the document]
Comment (victor-eds) [reply test/TritonGEN/tritongen-to-llvm.mlir]: Done!

File: test/TritonGEN/tritongen-to-llvm.mlir (hunk 2 above)
[the remark this answers is not in the document]
Comment (victor-eds) [reply test/TritonGEN/tritongen-to-llvm.mlir]: That won't change anything (tried), as we're creating new constants. We'd need to run an additional canonicalization pass, but that may mess up with other ops, so I'd keep as is.

**Comment by victor-eds** [comment]
> Just test case nits; code LGTM. You probably have to watch out for #1259 which extends the name mangling for the OCL builtins (`transform`/`transpose`).

I think this should work regardless of that. We would just accept more cases.

File: test/TritonGEN/tritongen-to-llvm.mlir
Comment (whitneywhtsang) [inline test/TritonGEN/tritongen-to-llvm.mlir:1]: https://github.com/intel/intel-xpu-backend-for-triton/commit/c99dc29eddb115c3071d3f2dac87d7fc2fa0c9c1 created `tritongen-2Dblockload-to-llvm.mlir`. Please move the new cache control test cases there.

File: test/TritonGEN/tritongen-to-llvm.mlir [hunk 4]
```diff
@@ -297,16 +297,152 @@ llvm.func @triton_gen.dpas.f32(%c : vector<8xf32>, %a : vector<4xf32>, %b : vect
...
+  // CHECK:     llvm.func @triton_gen.2Dblockload(%arg0: !llvm.ptr<1>, %arg1: i32, %arg2: i32, %arg3: i32, %arg4: i32, %arg5: i32) {
+  // CHECK:  [[C16:%.*]] = llvm.mlir.constant(16 : i32) : i32
+  // CHECK-NEXT:  [[DEST:%.*]] = llvm.alloca [[C16]] x i16 : (i32) -> !llvm.ptr
+  // CHECK-DAG:  [[ZERO:%.*]] = llvm.mlir.constant(0 : i32) : i32
+  // CHECK-DAG:  [[ONE:%.*]] = llvm.mlir.constant(1 : i32) : i32
+  // CHECK-DAG:  [[UNDEF:%.*]] = llvm.mlir.undef : vector<2xi32>
+  // CHECK-NEXT: [[COORD0:%.*]] = llvm.insertelement %arg4, [[UNDEF]][[[ZERO]] : i32] : vector<2xi32>
+  // CHECK-NEXT: [[COORD1:%.*]] = llvm.insertelement %arg5, [[COORD0]][[[ONE]] : i32] : vector<2xi32>
+  // CHECK-NEXT: llvm.call spir_funccc @_Z40intel_sub_group_2d_block_read_8b_8r32x2cPU3AS1viiiDv2_iPt(%arg0, %arg1, %arg2, %arg3, [[COORD1]], [[DEST]]) {passthrough = ["convergent"], triton_gen.DecorationCacheControlINTEL = #triton_gen.decoration_cache_control<#triton_gen.load_cache_control<1, Uncached, 0>, #triton_gen.load_cache_control<3, Cached, 0>>} : (!llvm.ptr<1>, i32, i32, i32, vector<2xi32>, !llvm.ptr) -> ()
+  // CHECK-NEXT: llvm.load [[DEST]] : !llvm.ptr -> vector<16xi16>
```
Comment (whitneywhtsang) [inline test/TritonGEN/tritongen-to-llvm.mlir:331]: Let's only test the full codegen in the first test.
```suggestion
llvm.func @triton_gen.2Dblockload(%ptr : !llvm.ptr<1>, %base_width : i32, %base_height : i32, %base_pitch : i32, %x : i32, %y : i32) {
  // CHECK: llvm.call spir_funccc @_Z40intel_sub_group_2d_block_read_8b_8r32x2cPU3AS1viiiDv2_iPt(%arg0, %arg1, %arg2, %arg3, [[COORD1]], [[DEST:%.*]]) {passthrough = ["convergent"], triton_gen.DecorationCacheControlINTEL = #triton_gen.decoration_cache_control<#triton_gen.load_cache_control<1, Uncached, 0>, #triton_gen.load_cache_control<3, Cached, 0>>} : (!llvm.ptr<1>, i32, i32, i32, vector<2xi32>, !llvm.ptr) -> ()
  // CHECK-NEXT: llvm.load [[DEST]] : !llvm.ptr -> vector<16xi16>
```

**Review by whitneywhtsang — APPROVED** [review]

**Comment by whitneywhtsang** [comment]
This change looks good to me, let's merge this after verifying it works end to end.

**Pull request merged**